qiskit_experiments/library/randomized_benchmarking/clifford_utils.py

```diff
@@ -41,14 +42,114 @@
+    for inst in circuit:
+        if inst.operation.name.startswith("Clifford"):  # Decompose
+            rule = inst.operation.definition.data
+            if len(rule) == 1 and len(inst.qubits) == len(rule[0].qubits):
+                if inst.operation.definition.global_phase:
+                    res.global_phase += inst.operation.definition.global_phase
+                res._data.append(
+                    CircuitInstruction(
+                        operation=rule[0].operation,
+                        qubits=inst.qubits,
+                        clbits=inst.clbits,
+                    )
+                )
+            else:
+                _circuit_compose(res, inst.operation.definition, qubits=inst.qubits)
+        else:  # Keep the original instruction
+            res._data.append(inst)
+    return res
+
+
+def _apply_qubit_layout(circuit: QuantumCircuit, layout: Sequence[int]) -> QuantumCircuit:
+    # Mapping qubits in circuit to physical qubits (layout)
+    res = QuantumCircuit(1 + max(layout), name=circuit.name, metadata=circuit.metadata)
+    res.add_bits(circuit.clbits)
+    for reg in circuit.cregs:
+        res.add_register(reg)
+    _circuit_compose(res, circuit, qubits=layout)
+    res._parameter_table = circuit._parameter_table
+    return res
+
+
+def _circuit_compose(
+    self: QuantumCircuit, other: QuantumCircuit, qubits: Sequence[Union[Qubit, int]]
+) -> QuantumCircuit:
+    # Simplified QuantumCircuit.compose with clbits=None, front=False, inplace=True, wrap=False
+    # without any validation, parameter_table/calibrations updates and copy of operations
+    # The input circuit `self` is changed inplace.
+    qubit_map = {
+        other.qubits[i]: (self.qubits[q] if isinstance(q, int) else q) for i, q in enumerate(qubits)
+    }
+    for instr in other:
+        self._data.append(
+            CircuitInstruction(
+                operation=instr.operation,
+                qubits=[qubit_map[q] for q in instr.qubits],
+                clbits=instr.clbits,
+            ),
+        )
+    self.global_phase += other.global_phase
+    return self
+
+
+def _truncate_inactive_qubits(
+    circ: QuantumCircuit, active_qubits: Sequence[Qubit]
+) -> QuantumCircuit:
+    new_data = []
+    for inst in circ:
+        if all(q in active_qubits for q in inst.qubits):
+            new_data.append(inst)
+
+    res = QuantumCircuit(active_qubits, name=circ.name)
+    res._calibrations = circ.calibrations
+    res._data = new_data
+    res._metadata = circ.metadata
+    return res
+
+
+# TODO: Naming: transform? translate? synthesis?
+def _transform_clifford_circuit(circuit: QuantumCircuit, basis_gates: Tuple[str]) -> QuantumCircuit:
+    # The function that synthesis clifford circuits with given basis gates,
+    # which should be commonly used during custom transpilation in the RB circuit generation.
+    return transpile(circuit, basis_gates=list(basis_gates), optimization_level=0)
```

Comment:
I am not sure about the purpose of this method. Note that you are using `optimization_level=0`, whereas I use `optimization_level=1` in the places where I call `transpile`. So this is not consistent.

Reply:
This function is intended to consolidate the code for basis translation in one place to facilitate changes to the specification around basis translation.

I think `optimization_level=0` is consistent with the default optimization level set in `transpile_options` of RB experiments (i.e. 0). I just want to keep the original spec in order to focus on the performance issue in this feature branch.
Could you tell us in the issue #922 if you find any good reason we should change `optimization_level=1` to be default?

Reply:
See my general comment below.